File: python/98_validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS with Recursion
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def recursive(root, low=-float('inf'), high=float('inf')):
            if not root:
                return True

            if low >= root.val or root.val >= high:
                return False
            if not recursive(root.left, low, root.val):
                return False
            if not recursive(root.right, root.val, high):
                return False
            return True

        return recursive(root)


#---------------------------------------------------------------------------------------#


# DFS with Stack
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True

        stack = [(root, -float('inf'), float('inf'))]

        while stack:
            root, low, high = stack.pop()
            if not root:
                continue

            if low >= root.val or root.val >= high:
                print('test')
                return False
            stack.append((root.left, low, root.val))
            stack.append((root.right, root.val, high))

        return True

File: go/98_validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return recursion(root, nil, nil)
}

func recursion(node *TreeNode, floor *int, ceil *int) bool {
	if node == nil {
		return true
	}
	if floor != nil && node.Val <= *floor {
		return false
	}
	if ceil != nil && node.Val >= *ceil {
		return false
	}
	return recursion(node.Left, floor, &node.Val) && recursion(node.Right, &node.Val, ceil)
}